docs/source/topics/websockets.rst

```diff
@@ -0,0 +1,1081 @@
+Websockets
+==========
+
+.. warning::
+
+  Websockets are considered an experimental API.  You'll need to opt-in
+  to this feature using the ``WEBSOCKETS`` feature flag:
+
+  .. code-block:: python
+
+    app = Chalice('myapp')
+    app.experimental_feature_flags.extend([
+        'WEBSOCKETS'
+    ])
+
+  See :doc:`experimental` for more information.
+
+
+
+Chalice supports websockets through integration with an API Gateway Websocket
+API. In a Chalice app
+
+Echo Server Example
+-------------------
+
+Below is an example of a simple echo server written with Chalice.
+
+.. code-block:: text
+   :caption: requirements.txt
+
+   boto3>=1.9.91
+
+
+.. code-block:: python
+   :caption: app.py
+   :linenos:
+
+   from boto3.session import Session
+
+   from chalice import Chalice
+   from chalice import WebsocketDisconnectedError
+
+   app = Chalice(app_name="echo-server")
+   app.websocket_api.session = Session()
+   app.experimental_feature_flags.update([
+       'WEBSOCKETS'
+   ])
+
+
+   @app.on_ws_message()
+   def message(event):
+   try:
+       app.websocket_api.send(
+	   connection_id=event.connection_id,
+	   message=event.body,
+       )
+   except WebsocketDisconnectedError as e:
+       pass  # Disconnected so we can't send the message back.
+
+
+Stepping through this app line by line, the first thing to note is that we
+need to import and instantiate a boto3 session. This session is manually
+assigned to ``app.websocket_api.session``.
+This is needed because in order to send websocket responses to API Gateway we
+need to construct a boto3 client. Chalice does not take a direct dependency
+on boto3 or botocore, so we need to provide the Session ourselves.
+
+.. code-block:: python
+
+   from boto3.session import Session
+   app.websocket_api.session = Session()
```

Comment:
I'm a little hesitant to use the boto3 session directly. Mainly because it does not suit well for users that want to have more granular control over how the client gets instantiated and how the message gets sent back. I think it would be worth exploring how we can have a builtin `WebsocketSession` that you could subclass from or provide your own session to. I will get back to providing more ideas for this once I've had a deeper look at the code.

Reply:
Chatted offline about this.  The thing that I was hoping to improve was to unify the configuration of the `WebsocketAPI` because you have to do both `websocket_api.session = Session()` as well as `websocket_api.configure(..., ...)` to get the object to a point where `send()` will work correctly.  I get why it's two steps, the `configure()` call actually happens in the chalice runtime, but it would be nice to consolidate.

However, there's the more general question about whether we should use boto3 session's directly.  My take on it is that it seems like a reasonable middle ground for the time being, but I think you're right about people wanting more control over the clients and how they get instantiated.  Either that or they want us to take care of everything.

John had an interesting idea about how chalice could have better automatic support with boto3.  We don't include it right now because it's a large dependency, especially if you don't need it.  A default chalice app is a couple of small .py files, which is nice.  However, what if we know you *do* need it?

The chalice mentality is to do things for the user they we know we need to do, but let the override it if they want (e.g. policy generation).  So one idea (and it's just an idea for now) is to detect if the user needs boto3, and if they do, to just _make it work_.  For example, if they have websocket routes defined, and they're calling `websock_api.send()`, then you _must_ have a boto3 client to do that.  So when we detect that we can check:

* Is boto3 in the deployment package?  If not, automatically include it for them.
* Is there a boto3 `_client` on the websocket api?  If not create one for them.

At any point the user can override this (add boto3 to their requirements.txt file), or manually set a client on the websocket API, but by default we can just try to do the right thing for them.  Obviously there's a lot of details and edge cases to work out, but I think it's a really interesting idea.

For the meantime, given this is an experimental API, as long as we clearly document that this API might change I'm ok with keeping this for now, but I'm interested to hear any alternatives if you have them.